Test plan followup

333fred opened the pull request:
Added support for the remaining compiler-side required test plan work. @AlekseyTs @dotnet/roslyn-compiler for review. I recommend reviewing this PR commit-by-commit for ease of review.

Relates to https://github.com/dotnet/roslyn/issues/43321

333fred commented:
@AlekseyTs @dotnet/roslyn-compiler for review.

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1364,15 +1363,31 @@ IAssemblySymbol computeContainingAssembly(ISymbol s)
+                            return assemblyIsInReferences(s.ContainingAssembly ?? this.Assembly);
                         default:
-                            return s.ContainingAssembly;
+                            return assemblyIsInReferences(s.ContainingAssembly);
```
>return assemblyIsInReferences(s.ContainingAssembly); [](start = 28, length = 52)

It looks like for a NamedType which isn't a definition we should check all type arguments, including from containing types. Could be a breaking change though.

333fred replied:
Filed [https://github.com/dotnet/roslyn/issues/44345](https://github.com/dotnet/roslyn/issues/44345) to follow up on this.

---
In reply to: [426081997](https://github.com/dotnet/roslyn/pull/44257#discussion_r426081997) [](ancestors = 426081997)

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1364,15 +1363,31 @@ IAssemblySymbol computeContainingAssembly(ISymbol s)
+                            var funcPtr = (IFunctionPointerTypeSymbol)s;
+                            if (!isContainingAssemblyInReferences(funcPtr.Signature.ReturnType))
+                            {
+                                return false;
```
>false [](start = 39, length = 5)

Is this code path covered by tests?

333fred replied:
Added a new test to `AccessCheckTests` for these.

---
In reply to: [426082638](https://github.com/dotnet/roslyn/pull/44257#discussion_r426082638) [](ancestors = 426082638)

AlekseyTs commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -1364,15 +1363,31 @@ IAssemblySymbol computeContainingAssembly(ISymbol s)
+                            {
+                                if (!isContainingAssemblyInReferences(param.Type))
+                                {
+                                    return false;
```
>return false; [](start = 36, length = 13)

Is this code path covered by tests?

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs:
```diff
@@ -25,6 +25,7 @@ internal enum TypeCompareKind
         IgnoreNullableModifiersForReferenceTypes = 8,
         ObliviousNullableModifierMatchesAny = 16,
         IgnoreNativeIntegers = 32,
+        FunctionPointerRefMatchesOutInRefReadonly = 64,
```
>FunctionPointerRefMatchesOutInRefReadonly [](start = 8, length = 41)

Consider adding a doc comment, including a brief explanation when is this needed.

AlekseyTs replied:
Also, it feels strange to have an "ignore" option that is not part of ```AllIgnoreOptions```. Might want to explain this as well.

---
In reply to: [426085624](https://github.com/dotnet/roslyn/pull/44257#discussion_r426085624) [](ancestors = 426085624)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs:
```diff
@@ -313,6 +313,10 @@ internal class MemberSignatureComparer : IEqualityComparer<Symbol>
             _typeComparison = typeComparison;
+            if (!considerRefKindDifferences)
+            {
+                _typeComparison |= TypeCompareKind.FunctionPointerRefMatchesOutInRefReadonly;
```
>TypeCompareKind.FunctionPointerRefMatchesOutInRefReadonly [](start = 35, length = 57)

Can we assert that this option is never set on entry?

333fred replied:
We could, but I'm not sure what invariant that's would be trying to preserve? Could you elaborate on what type of bug you think that would catch?

---
In reply to: [426087571](https://github.com/dotnet/roslyn/pull/44257#discussion_r426087571) [](ancestors = 426087571)

AlekseyTs replied:
> We could, but I'm not sure what invariant that would be trying to preserve? Could you elaborate on what type of bug you think that would catch?

From reviewing the change, I assumed that ```TypeCompareKind.FunctionPointerRefMatchesOutInRefReadonly``` is a very special flag. It is meant to be used by this component and it is really not appropriate (suspicious) to use it anywhere else in the code base. If the assumption is correct, the assert would explicitly indicate the intent and could possibly prevent people from misusing the flag. 

---
In reply to: [426889306](https://github.com/dotnet/roslyn/pull/44257#discussion_r426889306) [](ancestors = 426889306,426087571)

AlekseyTs commented:
>             => Hash.Combine(ReturnType, Hash.Combine(CallingConvention.GetHashCode(), RefKind.GetHashCode()));

It feels like this should be relaxed, since we are relaxing RefKind comparison.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs:400 in 447d87f. [](commit_id = 447d87f6a0f43786ac15d53f3d0ff20a8c53f75f, deletion_comment = False)

AlekseyTs commented:
>             => Hash.Combine(TypeWithAnnotations.GetHashCode(), RefKind.GetHashCode());

It feels like this should be relaxed, since we are relaxing RefKind comparison.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerParameterSymbol.cs:63 in 447d87f. [](commit_id = 447d87f6a0f43786ac15d53f3d0ff20a8c53f75f, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -157,5 +157,22 @@ internal override TypeSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnota
+        {
+            if ((compareKind & TypeCompareKind.FunctionPointerRefMatchesOutInRefReadonly) != 0)
+            {
+                return (refKind1, refKind2) switch
```
>(refKind1, refKind2) switch [](start = 23, length = 27)

Why so complicated? It looks like MemberSignatureComparer simply does  ```(refKind1 == RefKind.None) == (refKind2 == RefKind.None)```

333fred replied:
I found this significantly more readable than the version in `MemberSignatureComparer`.

---
In reply to: [426089324](https://github.com/dotnet/roslyn/pull/44257#discussion_r426089324) [](ancestors = 426089324)

AlekseyTs replied:
> I found this significantly more readable than the version in MemberSignatureComparer.

I spent significant amount of time trying to figure out your intent here (time I could have spent elsewhere) and whether all the right combinations are handled and handled properly. Every reader will have to go through the same trouble over and over again. I understand your desire to use switch expressions everywhere you can, but in this case, I believe it makes code less clear and unnecessarily complicated. Please use simplified form of the check, especially that we are already using it elsewhere. 

---
In reply to: [426821892](https://github.com/dotnet/roslyn/pull/44257#discussion_r426821892) [](ancestors = 426821892,426089324)

333fred replied:
>  I understand your desire to use switch expressions everywhere you can

To be clear, that is not the motivation.

> I spent significant amount of time trying to figure out your intent here (time I could have spent elsewhere) and whether all the right combinations are handled and handled properly. 

And _I_ spent probably a similar amount of time creating a truth table in my head for the original check to be sure that I could understand what it was checking. I used this form because _I_ genuinely find it significantly easier to read. I don't find the other from simpler in any fashion.

---
In reply to: [426825848](https://github.com/dotnet/roslyn/pull/44257#discussion_r426825848) [](ancestors = 426825848,426821892,426089324)

333fred replied:
I'll change this. However, I want to be clear that I don't just use new features for the fun of it. I use them when it would help me to read the code better, and I find that writing out the truth table (as I can do concisely here with a switch expression) is significantly easier to understand. I'll change this because we have prior art in the form of a comparison, but in future code that does similar things I'm still going to prefer using the form I understand more easily.

---
In reply to: [426840351](https://github.com/dotnet/roslyn/pull/44257#discussion_r426840351) [](ancestors = 426840351,426825848,426821892,426089324)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/CustomModifierUtils.cs:
```diff
@@ -102,7 +102,8 @@ internal static TypeSymbol CopyTypeCustomModifiers(TypeSymbol sourceType, TypeSy
 
-            Debug.Assert(resultType.Equals(destinationType, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds)); // Same object/dynamic, nullability and tuple names as destination type.
+            // Same object/dynamic, nullability and tuple names as destination type.
+            Debug.Assert(resultType.Equals(destinationType, TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds | TypeCompareKind.IgnoreNativeIntegers));
```
> | TypeCompareKind.IgnoreNativeIntegers [](start = 124, length = 39)

This doesn't feel right. Native int types specified in source should be preserved.

333fred replied:
@cston can you comment? Regardless I don't want to make a change around native int encoding in this PR, so I'm either going to skip the tests that hit this assert or keep this as is.

---
In reply to: [426089925](https://github.com/dotnet/roslyn/pull/44257#discussion_r426089925) [](ancestors = 426089925)

cston replied:
It sounds like a bug. Please log an issue, thanks.

---
In reply to: [426823520](https://github.com/dotnet/roslyn/pull/44257#discussion_r426823520) [](ancestors = 426823520,426089925)

333fred replied:
Filed [https://github.com/dotnet/roslyn/issues/44358](https://github.com/dotnet/roslyn/issues/44358).

---
In reply to: [426836636](https://github.com/dotnet/roslyn/pull/44257#discussion_r426836636) [](ancestors = 426836636,426823520,426089925)

333fred replied:
Removed the flag and skipped the relevant test.

---
In reply to: [426894790](https://github.com/dotnet/roslyn/pull/44257#discussion_r426894790) [](ancestors = 426894790,426836636,426823520,426089925)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5571,6 +5571,915 @@ .maxstack  1
+}}
+unsafe class Derived : Base
+{{
+    protected override void M1(delegate*<{type2}> ptr) {{}}
```
>M1 [](start = 28, length = 2)

Consider asserting what signature do we end up with at the end.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5571,6 +5571,915 @@ .maxstack  1
+}
+unsafe class Derived : Base
+{
+    protected override void M1(delegate*<(int i, string s)> ptr) {{}}
```
>M1 [](start = 28, length = 2)

Consider asserting what signature do we end up with at the end.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5571,6 +5571,915 @@ .maxstack  1
+}
+unsafe class Derived : Base
+{
+    protected override void M1(delegate*<string> ptr) {{}}
```
>M1 [](start = 28, length = 2)

Consider asserting what signature do we end up with at the end.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5571,6 +5571,915 @@ .maxstack  1
+}
+unsafe class Derived : Base
+{
+    protected override void M1(delegate*<ref string> ptr) {{}}
```
>M1 [](start = 28, length = 2)

Consider asserting what signature do we end up with at the end.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5571,6 +5571,915 @@ .maxstack  1
+}
+unsafe class Derived : Base
+{
+    protected override void M1(ref delegate*<string> ptr) {{}}
```
>M1 [](start = 28, length = 2)

Consider asserting what signature do we end up with at the end.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -1381,20 +1381,25 @@ void M(bool b)
                 // (13,13): error CS0173: Type of conditional expression cannot be determined because there is no implicit conversion between 'delegate*<string>' and 'delegate*<string?>'
                 //         _ = b ? ptr1 : ptr4;
                 Diagnostic(ErrorCode.ERR_InvalidQM, "b ? ptr1 : ptr4").WithArguments("delegate*<string>", "delegate*<string?>").WithLocation(13, 13),
+                // (14,24): warning CS8619: Nullability of reference types in value of type 'delegate*<string?>' doesn't match target type 'delegate*<string>'.
```
>delegate*<string> [](start = 140, length = 17)

Shouldn't we infer ```delegate*<string?>``` instead?

333fred replied:
No, these are `ref`s. Unfortunately, our CSharpErrorMessageFormat does not include return ref kinds, so the error message isn't super clear. I could change that to include them, but I'm unsure of the fully ramifications of such a change.

---
In reply to: [426093508](https://github.com/dotnet/roslyn/pull/44257#discussion_r426093508) [](ancestors = 426093508)

AlekseyTs replied:
> Unfortunately, our CSharpErrorMessageFormat does not include return ref kinds, so the error message isn't super clear. 

Since this isn't really a member, perhaps the symbol display for function pointers shouldn't be driven by the options for members and should always display ```ref``` kind because this is important aspect of the type. Same for other things we might be dropping  

---
In reply to: [426844096](https://github.com/dotnet/roslyn/pull/44257#discussion_r426844096) [](ancestors = 426844096,426093508)

333fred replied:
I've added a note to the tracking issue to consider that change, will resolve this thread.

---
In reply to: [426910541](https://github.com/dotnet/roslyn/pull/44257#discussion_r426910541) [](ancestors = 426910541,426844096,426093508)

AlekseyTs commented:
>         public static bool HaveSameReturnTypes(MethodSymbol member1, MethodSymbol member2, TypeCompareKind typeComparison)

Are all consumers passing the right ```TypeCompareKind``` for function pointers?

---
Refers to: src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs:463 in 3d33fe4. [](commit_id = 3d33fe415f68352bbbdc9bee32e83625546d67ab, deletion_comment = False)

AlekseyTs commented:
Done with review pass (iteration 7)

333fred commented:
>         public static bool HaveSameReturnTypes(MethodSymbol member1, MethodSymbol member2, TypeCompareKind typeComparison)

We could actually remove this API. There are no consumers.

---
In reply to: [629557451](https://github.com/dotnet/roslyn/pull/44257#issuecomment-629557451) [](ancestors = 629557451)

---
Refers to: src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs:463 in 3d33fe4. [](commit_id = 3d33fe415f68352bbbdc9bee32e83625546d67ab, deletion_comment = False)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [629557943](https://github.com/dotnet/roslyn/pull/44257#issuecomment-629557943) [](ancestors = 629557943)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerTypeSymbol.cs:
```diff
@@ -157,5 +157,29 @@ internal override TypeSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnota
+        /// for types that only differ by the type of ref they have.
+        /// </summary>
+        internal static RefKind GetRefKindForHashCode(RefKind refKind)
+            => refKind switch
```
>refKind switch [](start = 15, length = 14)

I think it would be more robust and clearer to base the hash code on the value ```(refKind == RefKind.None)```. There would be clear correspondence to the equality evaluation, consistency between implementation of both APIs would be obvious and wouldn't require any additional maintenance when the set of RefKinds is extended.

AlekseyTs replied:
There is also an option to not adjust hash code based on ref kinds.

---
In reply to: [426986569](https://github.com/dotnet/roslyn/pull/44257#discussion_r426986569) [](ancestors = 426986569)

AlekseyTs commented:
Done with review pass (iteration 8)

333fred commented:
@AlekseyTs address feedback.

---
In reply to: [630534288](https://github.com/dotnet/roslyn/pull/44257#issuecomment-630534288) [](ancestors = 630534288)

333fred commented:
@dotnet/roslyn-compiler for a second review. I'm going to hold off on rebasing until reviews are done.

333fred commented:
@dotnet/roslyn-compiler for a second review please.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5571,6 +5571,1125 @@ .maxstack  1
+                Diagnostic(ErrorCode.ERR_CantChangeReturnTypeOnOverride, "M2").WithArguments("Derived.M2()", "Base.M2()", $"delegate*<{refKindString(refKind1)}string, void>").WithLocation(10, 48 + refKind2.Length)
+            );
+
+            static string refKindString(string orig) => orig == " " ? "" : orig;
```
>refKindString [](start = 26, length = 13)

Could this local function be removed if the `RefKind.None` cases were "" rather than " "?

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5571,6 +5571,1125 @@ .maxstack  1
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class Base
+{
+    protected virtual void M1(delegate*<(int, string)> ptr) {{}}
```
>{{}} [](start = 60, length = 4)

Is the inner `{}` needed? Same question for tests below.

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/AccessCheckTests.cs:
```diff
@@ -1478,5 +1502,26 @@ static void Main()
+
+            var b = comp2.GetMember("B").GetPublicSymbol();
+
+            Assert.Throws<ArgumentException>(() => ((Compilation)comp2).IsSymbolAccessibleWithin(ptr1, b));
```
>IsSymbolAccessibleWithin [](start = 72, length = 24)

Why does this throw `ArgumentException` rather than returning `false`?

333fred replied:
Because `A` isn't from any reference of `comp2`.

---
In reply to: [428179712](https://github.com/dotnet/roslyn/pull/44257#discussion_r428179712) [](ancestors = 428179712)

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -1381,20 +1381,25 @@ void M(bool b)
+                // (23,24): warning CS8619: Nullability of reference types in value of type 'delegate*<string?>' doesn't match target type 'delegate*<string>'.
+                //         _ = b ? ptr7 : ptr8;
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "ptr8").WithArguments("delegate*<string?>", "delegate*<string>").WithLocation(23, 24)
             );
```
>); [](start = 12, length = 2)

Are we testing mismatch of parameter types in addition to return types?

cston commented on src/Compilers/CSharp/Test/Semantic/Semantics/OverloadResolutionTests.cs:
```diff
@@ -843,6 +843,94 @@ .namespace System.Runtime.CompilerServices
+        public void NormalizeTaskTypes_FunctionPointers()
+        {
+
+            string source =
```
Blank line.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5571,6 +5571,1125 @@ .maxstack  1
+        public void Override_SingleDimensionArraySizesInMetadata()
+        {
+
+            var il = @"
```
Blank line.

cston commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -646,6 +646,37 @@ static void assertSymbolEquality(Symbol s1, Symbol s2, Equality eq)
+            Assert.Equal(ptr1Ref.GetHashCode(), ptr1RefReadonly.GetHashCode());
+            Assert.Equal(ptr2Ref.GetHashCode(), ptr2In.GetHashCode());
+            Assert.Equal(ptr2Ref.GetHashCode(), ptr2Out.GetHashCode());
+        }
```
>} [](start = 8, length = 1)

Are we testing `symbolEqualityComparer.GetHashCode()`?

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -5571,6 +5571,1122 @@ .maxstack  1
+
+            comp.VerifyDiagnostics(
+                // (16,29): warning CS8610: Nullability of reference types in type of parameter 'ptr' doesn't match overridden member.
+                //     protected override void M1(delegate*<string> ptr) {{}}
```
>{{}} [](start = 73, length = 4)

Minor: Comments are stale, here and below.

333fred commented:
Debug integration test failure is https://github.com/dotnet/roslyn/issues/43967.
